File: src/application/services/issue-application-service.ts

import { Issue } from '../../domain/entities/issue.js';
import { IssueId } from '../../domain/value-objects/issue-id.js';
import { IssueStatus } from '../../domain/value-objects/issue-status.js';
import { IssueType } from '../../domain/value-objects/issue-type.js';
import { ProjectId } from '../../domain/value-objects/project-id.js';
import { IssueDtoMapper } from '../dtos/issue-dto.js';

import type { TimeProvider } from '../../domain/interfaces/time-provider.js';
import type { IssueRepository } from '../../domain/repositories/issue-repository.js';
import type { ProjectRepository } from '../../domain/repositories/project-repository.js';
import type { UnitOfWork } from '../../domain/repositories/session-repository.js';
import type { 
  CreateIssueCommand, 
  UpdateIssueCommand, 
  IssueDto, 
  IssueOperationResult 
} from '../dtos/issue-dto.js';

/**
 * Application service for Issue-related operations
 * Orchestrates use cases and coordinates between domain entities and infrastructure
 */
export class IssueApplicationService {
  constructor(
    private readonly issueRepository: IssueRepository,
    private readonly projectRepository: ProjectRepository,
    private readonly unitOfWork: UnitOfWork,
    private readonly timeProvider: TimeProvider
  ) {}

  /**
   * Create a new issue with hierarchy and estimation validation
   */
  async createIssue(command: CreateIssueCommand): Promise<IssueOperationResult> {
    try {
      return await this.unitOfWork.execute(async () => {
        const validation = await this.validateCreateIssueCommand(command);

        if (!validation.isValid) {
          return this.createErrorResult(validation.error!);
        }

        const { projectId, parentId } = validation;
        const issue = this.createDomainIssue(command, projectId!);

        if (parentId) {
          await this.setIssueParent(issue, parentId);
        }

        await this.persistIssue(issue, projectId!);

        return this.createSuccessResult(issue);
      });
    } catch (error) {
      return this.createErrorResult(error instanceof Error ? error.message : 'Unknown error occurred');
    }
  }

  /**
   * Update an existing issue with validation
   */
  async updateIssue(command: UpdateIssueCommand): Promise<IssueOperationResult> {
    try {
      return await this.unitOfWork.execute(async () => {
        const issue = await this.findIssueById(command.id);

        if (!issue) {
          return this.createErrorResult('Issue not found');
        }

        const updateResult = this.applyIssueUpdates(issue, command);

        if (!updateResult.isValid) {
          return this.createErrorResult(updateResult.error!);
        }

        await this.issueRepository.save(issue);

        return this.createSuccessResult(issue);
      });
    } catch (error) {
      return this.createErrorResult(error instanceof Error ? error.message : 'Unknown error occurred');
    }
  }

  /**
   * Get an issue by ID
   */
  async getIssue(issueId: string): Promise<IssueDto | null> {
    try {
      if (!issueId) {
        throw new Error('Issue ID is required');
      }

      const id = IssueId.from(issueId);
      const issue = await this.issueRepository.findById(id);
      
      return issue ? IssueDtoMapper.toDto(issue) : null;
    } catch {
      // Handle repository errors gracefully
      return null;
    }
  }

  /**
   * Get all issues for a project
   */
  async getProjectIssues(projectId: string): Promise<IssueDto[]> {
    if (!projectId) {
      throw new Error('Project ID is required');
    }

    try {
      const id = ProjectId.from(projectId);
      const issues = await this.issueRepository.findByProjectId(id);
      
      return issues.map(issue => IssueDtoMapper.toDto(issue));
    } catch {
      // Handle repository errors gracefully
      return [];
    }
  }

  /**
   * Validate issue hierarchy rules
   */
  private async validateIssueHierarchy(
    type: string, 
    parentId?: string
  ): Promise<{ isValid: boolean; error?: string }> {
    // Epic cannot have parent
    if (type === IssueType.Epic && parentId) {
      return {
        isValid: false,
        error: 'Epic cannot have a parent'
      };
    }

    // Subtask must have parent
    if (type === IssueType.Subtask && !parentId) {
      return {
        isValid: false,
        error: 'Subtask must have a parent'
      };
    }

    // If parent is specified, validate parent-child relationship
    if (parentId) {
      const parentIssueId = IssueId.from(parentId);
      const parent = await this.issueRepository.findById(parentIssueId);
      
      if (!parent) {
        return {
          isValid: false,
          error: 'Parent issue not found'
        };
      }

      // Subtask can only have Story parent
      if (type === IssueType.Subtask && parent.type !== IssueType.Story) {
        return {
          isValid: false,
          error: 'Subtask must have a Story parent'
        };
      }

      // Story cannot have Subtask parent
      if (type === IssueType.Story && parent.type === IssueType.Subtask) {
        return {
          isValid: false,
          error: 'Story cannot have Subtask parent'
        };
      }
    }

    return { isValid: true };
  }

  /**
   * Validate estimation rules
   */
  private validateEstimationRules(
    type: string, 
    estimate?: number
  ): { isValid: boolean; error?: string } {
    // Epic cannot have estimate
    if (type === IssueType.Epic && estimate !== undefined) {
      return {
        isValid: false,
        error: 'Epic cannot have estimate'
      };
    }

    // If estimate is provided, validate it's a valid Fibonacci number
    if (estimate !== undefined) {
      const fibonacciSequence = [1, 2, 3, 5, 8, 13, 21, 34, 55, 89];

      if (!fibonacciSequence.includes(estimate)) {
        return {
          isValid: false,
          error: 'Estimate must follow Fibonacci sequence (1, 2, 3, 5, 8, 13, etc.)'
        };
      }
    }

    return { isValid: true };
  }

  /**
   * Validate create issue command comprehensively
   */
  private async validateCreateIssueCommand(
    command: CreateIssueCommand
  ): Promise<{ isValid: boolean; error?: string; projectId?: ProjectId; parentId?: IssueId }> {
    // Validate required fields
    if (!command.title?.trim()) {
      return { isValid: false, error: 'Issue title is required' };
    }

    if (!command.projectId) {
      return { isValid: false, error: 'Project ID is required' };
    }

    // Validate project exists
    const projectId = ProjectId.from(command.projectId);
    const project = await this.projectRepository.findById(projectId);

    if (!project) {
      return { isValid: false, error: 'Project not found' };
    }

    // Validate hierarchy rules
    const hierarchyValidation = await this.validateIssueHierarchy(command.type, command.parentId);

    if (!hierarchyValidation.isValid) {
      return { isValid: false, error: hierarchyValidation.error || 'Hierarchy validation failed' };
    }

    // Validate estimation rules
    const estimationValidation = this.validateEstimationRules(command.type, command.estimate);

    if (!estimationValidation.isValid) {
      return { isValid: false, error: estimationValidation.error || 'Estimation validation failed' };
    }

    const result: { isValid: boolean; error?: string; projectId?: ProjectId; parentId?: IssueId } = { isValid: true, projectId };
    
    if (command.parentId) {
      result.parentId = IssueId.from(command.parentId);
    }
    
    return result;
  }

  /**
   * Create domain issue entity from command
   */
  private createDomainIssue(command: CreateIssueCommand, projectId: ProjectId): Issue {
    const issue = Issue.create(
      command.title.trim(),
      command.description || '',
      command.type,
      this.timeProvider,
      projectId
    );

    if (command.estimate !== undefined) {
      issue.setEstimate(command.estimate);
    }

    return issue;
  }

  /**
   * Set parent relationship for issue
   */
  private async setIssueParent(issue: Issue, parentId: IssueId): Promise<void> {
    issue.setParent(parentId);
    
    // Update parent's children list
    const parent = await this.issueRepository.findById(parentId);

    if (parent) {
      parent.addChild(issue.id);
      await this.issueRepository.save(parent);
    }
  }

  /**
   * Persist issue to repository
   */
  private async persistIssue(issue: Issue, projectId: ProjectId): Promise<void> {
    await this.issueRepository.saveToProject(issue, projectId);
  }

  /**
   * Find issue by ID safely
   */
  private async findIssueById(issueId: string): Promise<Issue | null> {
    const id = IssueId.from(issueId);

    return await this.issueRepository.findById(id);
  }

  /**
   * Apply updates to an issue with validation
   */
  private applyIssueUpdates(
    issue: Issue, 
    command: UpdateIssueCommand
  ): { isValid: boolean; error?: string } {
    // Update title if provided
    if (command.title !== undefined) {
      if (!command.title.trim()) {
        return { isValid: false, error: 'Issue title cannot be empty' };
      }
      issue.updateTitle(command.title.trim());
    }

    // Update description if provided
    if (command.description !== undefined) {
      issue.updateDescription(command.description);
    }

    // Update status if provided
    if (command.status !== undefined) {
      if (!IssueStatus.canTransition(issue.status, command.status)) {
        return { 
          isValid: false, 
          error: `Invalid status transition from ${issue.status} to ${command.status}` 
        };
      }
      issue.updateStatus(command.status);
    }

    // Update estimate if provided
    if (command.estimate !== undefined) {
      const estimationValidation = this.validateEstimationRules(issue.type, command.estimate);

      if (!estimationValidation.isValid) {
        return { isValid: false, error: estimationValidation.error || 'Estimation validation failed' };
      }
      issue.setEstimate(command.estimate);
    }

    return { isValid: true };
  }

  /**
   * Create error result
   */
  private createErrorResult(error: string): IssueOperationResult {
    return { success: false, error };
  }

  /**
   * Create success result
   */
  private createSuccessResult(issue: Issue): IssueOperationResult {
    return { success: true, data: IssueDtoMapper.toDto(issue) };
  }
}

File: src/main/kotlin/io/spiralhouse/cycletime/application/services/IssueApplicationService.kt
package io.spiralhouse.cycletime.application.services

import io.spiralhouse.cycletime.application.commands.*
import io.spiralhouse.cycletime.application.dto.*
import io.spiralhouse.cycletime.application.exceptions.*
import io.spiralhouse.cycletime.domain.entities.Issue
import io.spiralhouse.cycletime.domain.exceptions.DomainException
import io.spiralhouse.cycletime.domain.repositories.IssueRepository
import io.spiralhouse.cycletime.domain.repositories.ProjectRepository
import io.spiralhouse.cycletime.domain.repositories.UnitOfWork
import io.spiralhouse.cycletime.domain.services.TimeProvider
import io.spiralhouse.cycletime.domain.valueobjects.*
import io.spiralhouse.cycletime.infrastructure.persistence.queries.HierarchyQueries

/**
 * Application service for managing Issue operations.
 *
 * This service acts as the orchestration layer between the domain and infrastructure,
 * coordinating complex business operations while maintaining transactional integrity
 * through the Unit of Work pattern. It enforces business rules, handles cross-aggregate
 * operations, and translates between domain entities and DTOs.
 *
 * ## Responsibilities:
 * - **Use Case Orchestration**: Each public method represents a complete use case
 * - **Transaction Management**: Ensures all operations within a use case are atomic
 * - **Domain Coordination**: Orchestrates operations across Issue and Project aggregates
 * - **DTO Translation**: Converts between domain entities and application DTOs
 * - **Business Rule Enforcement**: Validates commands and enforces application-level rules
 * - **Hierarchy Management**: Validates and maintains issue hierarchy rules
 * - **Dependency Management**: Handles issue dependencies and prevents circular references
 *
 * ## Design Patterns:
 * - **Unit of Work**: All operations wrapped in transactional boundaries
 * - **Command Pattern**: Uses command objects for complex operations
 * - **Repository Pattern**: Delegates persistence to repository interfaces
 * - **DTO Pattern**: Never exposes domain entities directly
 *
 * ## Thread Safety:
 * All methods are suspend functions designed for concurrent access.
 * The Unit of Work ensures transactional isolation between operations.
 *
 * @property issueRepository Repository for Issue aggregate persistence
 * @property projectRepository Repository for Project aggregate persistence
 * @property unitOfWork Transaction coordinator ensuring atomic operations
 * @property timeProvider Time provider for consistent temporal operations
 *
 * @constructor Creates a new IssueApplicationService with required dependencies
 */
class IssueApplicationService(
    private val issueRepository: IssueRepository,
    private val projectRepository: ProjectRepository,
    private val unitOfWork: UnitOfWork,
    private val timeProvider: TimeProvider
) : ApplicationServiceBase() {

    // ================================================================================
    // CRUD Operations
    // ================================================================================

    /**
     * Creates a new issue with the specified attributes.
     *
     * ## Business Rules:
     * - Issue title must be non-empty and <= 255 characters
     * - Description is optional but limited to 2000 characters if provided
     * - Hierarchy rules must be followed (Epic -> Story -> Subtask)
     * - Estimates required for Subtasks, forbidden for Epics
     * - Project inheritance from parent when parent has project
     *
     * @param command Command containing issue creation details
     * @return IssueDto representing the created issue with generated ID
     * @throws io.spiralhouse.cycletime.application.exceptions.HierarchyViolationException if hierarchy rules are violated
     * @throws io.spiralhouse.cycletime.application.exceptions.ProjectNotFoundException if project doesn't exist
     * @throws io.spiralhouse.cycletime.domain.exceptions.DomainException if domain rules are violated
     */
    suspend fun createIssue(command: CreateIssueCommand): IssueDto {
        return unitOfWork.execute {
            // Validate project exists if specified
            command.projectId?.let { projectId ->
                if (projectRepository.findById(projectId) == null) {
                    throw ProjectNotFoundException(projectId)
                }
            }

            // Validate parent and determine project inheritance
            val finalProjectId = when {
                command.parentId != null -> {
                    val parent = issueRepository.findById(command.parentId)
                        ?: throw IssueNotFoundException(command.parentId)

                    // Validate hierarchy rules
                    validateHierarchy(command.type, parent.type)

                    // Inherit project from parent if not explicitly specified
                    command.projectId ?: parent.projectId
                }
                else -> {
                    // No parent - validate types that require parent
                    validateHierarchy(command.type, null)
                    command.projectId
                }
            }

            // Create the issue
            val issue = Issue.create(
                title = command.title,
                description = command.description,
                type = command.type,
                parentId = command.parentId,
                projectId = finalProjectId,
                timeProvider = timeProvider
            )

            // Set estimate if provided
            command.estimate?.let { estimate ->
                // Delegate to domain entity for validation
                // Domain exceptions will propagate to caller
                issue.setEstimate(estimate)
            }

            // Set assignee if provided
            command.assigneeId?.let { assigneeId ->
                issue.updateAssignee(assigneeId)
            }

            issueRepository.save(issue)
            IssueDto.fromIssue(issue)
        }
    }

    /**
     * Retrieves an issue by its unique identifier.
     *
     * @param issueId The unique identifier of the issue to retrieve
     * @return IssueDto if found, null otherwise
     */
    suspend fun getIssue(issueId: IssueId): IssueDto? {
        return unitOfWork.execute {
            issueRepository.findById(issueId)?.let { issue ->
                IssueDto.fromIssue(issue)
            }
        }
    }

    /**
     * Updates an existing issue's attributes.
     *
     * @param command Command containing issue ID and optional updates
     * @return IssueDto representing the updated issue state
     * @throws io.spiralhouse.cycletime.application.exceptions.IssueNotFoundException if the issue doesn't exist
     * @throws io.spiralhouse.cycletime.domain.exceptions.DomainException if updates violate domain rules
     */
    suspend fun updateIssue(command: UpdateIssueCommand): IssueDto {
        return unitOfWork.execute {
            val issue = findIssueOrThrow(command.id)

            // Apply updates
            command.title?.let { title ->
                issue.updateTitle(title)
            }

            command.description?.let { description ->
                issue.updateDescription(description)
            }

            command.estimate?.let { estimate ->
                // Delegate to domain entity for validation
                // Domain exceptions will propagate to caller
                issue.setEstimate(estimate)
            }

            command.assigneeId?.let { assigneeId ->
                issue.updateAssignee(assigneeId)
            }

            issueRepository.save(issue)
            IssueDto.fromIssue(issue)
        }
    }

    /**
     * Soft-deletes an issue from the system.
     *
     * This operation marks the issue as deleted by setting the deleted_at timestamp
     * but preserves the data for potential recovery via restoreIssue().
     * Child issues are cascade soft-deleted (handled by repository).
     *
     * @param issueId The ID of the issue to delete
     * @throws io.spiralhouse.cycletime.application.exceptions.IssueNotFoundException if the issue doesn't exist
     */
    suspend fun deleteIssue(issueId: IssueId) {
        return unitOfWork.execute {
            val issue = findIssueOrThrow(issueId)
            issueRepository.softDelete(issueId)
        }
    }

    /**
     * Restores a soft-deleted issue.
     *
     * This operation clears the deleted_at timestamp, making the issue
     * active again. Note that child issues are NOT automatically restored -
     * they must be restored individually if needed.
     *
     * ## Business Rules:
     * - Issue must exist (including deleted issues) or IssueNotFoundException is thrown
     * - Operation is idempotent - restoring an already-active issue succeeds
     * - Child issues remain deleted (explicit restoration required)
     * - CRITICAL: Cannot restore child issue if parent is still deleted (must restore parent first)
     *
     * ## State Transitions:
     * - DELETED → ACTIVE (clears deleted_at)
     * - ACTIVE → ACTIVE (idempotent, no-op)
     *
     * @param id The ID of the issue to restore
     * @return IssueDto representing the restored issue
     * @throws IssueNotFoundException if the issue doesn't exist at all
     * @throws IllegalStateException if attempting to restore child when parent is deleted
     */
    suspend fun restoreIssue(id: IssueId): IssueDto {
        return unitOfWork.execute {
            val issue = issueRepository.findIncludingDeleted(id)
                ?: throw IssueNotFoundException(id)

            // Validate parent is not deleted
            if (issue.parentId != null) {
                // Check if parent exists (including deleted)
                val parentIncludingDeleted = issueRepository.findIncludingDeleted(issue.parentId)
                if (parentIncludingDeleted != null) {
                    // Parent exists - check if it's deleted by verifying findById returns null
                    val activeParent = issueRepository.findById(issue.parentId)
                    if (activeParent == null) {
                        // Parent exists in findIncludingDeleted but not in findById = deleted
                        throw IllegalStateException(
                            "Cannot restore issue ${issue.id.value} - parent ${parentIncludingDeleted.id.value} is still deleted. " +
                            "Restore parent first or contact support."
                        )
                    }
                }
            }

            // Restore issue (does NOT auto-restore children)
            issueRepository.restore(id)

            val restoredIssue = issueRepository.findById(id)
                ?: throw IssueNotFoundException(id)

            IssueDto.fromIssue(restoredIssue)
        }
    }

    // ================================================================================
    // Hierarchy Operations
    // ================================================================================

    /**
     * Moves an issue to a different parent.
     *
     * This operation validates hierarchy rules before moving and preserves
     * all issue attributes except the parent relationship. Project inheritance
     * is recalculated based on the new parent.
     *
     * ## Business Rules:
     * - Move must respect hierarchy rules (Epic->Story->Subtask)
     * - Cannot create circular parent relationships
     * - Project is inherited from new parent if parent has project
     * - All other attributes (status, dependencies, etc.) are preserved
     *
     * @param command Command containing issue and new parent IDs
     * @return IssueDto representing the moved issue with new parent
     * @throws HierarchyViolationException if move violates hierarchy rules
     * @throws IssueNotFoundException if issue or parent doesn't exist
     */
    suspend fun moveIssue(command: MoveIssueCommand): IssueDto {
        return unitOfWork.execute {
            val issue = findIssueOrThrow(command.issueId)

            // Determine new parent type and project
            val (newParentType, newProjectId) = when (command.newParentId) {
                null -> Pair(null, issue.projectId) // No parent - keep existing project
                else -> {
                    val parent = issueRepository.findById(command.newParentId)
                        ?: throw IssueNotFoundException(command.newParentId)

                    // Check for circular dependency (cannot move to own descendant)
                    if (isDescendantOf(parent.id, issue.id)) {
                        throw HierarchyViolationException(
                            issue.type,
                            parent.type,
                            "Cannot move issue to its own descendant"
                        )
                    }

                    Pair(parent.type, parent.projectId ?: issue.projectId)
                }
            }

            // Validate hierarchy rules for the move
            validateHierarchy(issue.type, newParentType)

            // Create issue with new parent while preserving all other attributes
            val movedIssue = Issue.fromSnapshot(
                issue.toSnapshot().copy(
                    parentId = command.newParentId,
                    projectId = newProjectId,
                    updatedAt = timeProvider.now()
                ),
                timeProvider
            )

            issueRepository.save(movedIssue)
            IssueDto.fromIssue(movedIssue)
        }
    }

    /**
     * Checks if an issue is a descendant of another issue.
     * Used to prevent circular parent relationships.
     *
     * @param potentialDescendantId The ID of the potential descendant
     * @param ancestorId The ID of the potential ancestor
     * @return true if potentialDescendantId is a descendant of ancestorId
     */
    private suspend fun isDescendantOf(potentialDescendantId: IssueId, ancestorId: IssueId): Boolean {
        var currentId: IssueId? = potentialDescendantId
        val visited = mutableSetOf<IssueId>()

        while (currentId != null) {
            if (currentId == ancestorId) return true
            if (!visited.add(currentId)) break // Prevent infinite loop

            val issue = issueRepository.findById(currentId) ?: break
            currentId = issue.parentId
        }

        return false
    }

    /**
     * Gets the complete hierarchy for an issue (including all descendants).
     *
     * @param issueId The root issue ID
     * @return IssueHierarchyDto representing the complete hierarchy
     * @throws io.spiralhouse.cycletime.application.exceptions.IssueNotFoundException if the issue doesn't exist
     */
    suspend fun getIssueHierarchy(issueId: IssueId): IssueHierarchyDto {
        return unitOfWork.execute {
            val issue = findIssueOrThrow(issueId)

            buildHierarchy(issue)
        }
    }

    /**
     * Gets the extended hierarchy information for an issue.
     * Returns the issue with its parent, direct children, and total descendant count.
     *
     * @param issueId The issue ID
     * @return IssueHierarchyExtendedDto representing the extended hierarchy
     * @throws io.spiralhouse.cycletime.application.exceptions.IssueNotFoundException if the issue doesn't exist
     */
    suspend fun getIssueHierarchyExtended(issueId: IssueId): IssueHierarchyExtendedDto {
        return unitOfWork.execute {
            val issue = findIssueOrThrow(issueId)

            val parent = issue.parentId?.let { parentId ->
                issueRepository.findById(parentId)
            }

            val children = issueRepository.findByParent(issue.id)
            // Use optimized query to avoid N+1 problem
            val totalDescendants = HierarchyQueries.countDescendantsOptimized(issue.id)

            IssueHierarchyExtendedDto.fromIssueWithParentAndChildren(
                issue = issue,
                parent = parent,
                children = children,
                totalDescendants = totalDescendants
            )
        }
    }

    private suspend fun buildHierarchy(issue: Issue): IssueHierarchyDto {
        val children = issueRepository.findByParent(issue.id)
        val childHierarchies = children.map { child -> buildHierarchy(child) }

        return IssueHierarchyDto.fromIssueWithChildren(
            issue = issue,
            childHierarchies = childHierarchies
        )
    }


    /**
     * Lists all direct children of a parent issue.
     *
     * @param parentId The parent issue ID
     * @return List of IssueDto representing direct children
     */
    suspend fun listChildren(parentId: IssueId): List<IssueDto> {
        return unitOfWork.execute {
            val children = issueRepository.findByParent(parentId)
            children.map { issue -> IssueDto.fromIssue(issue) }
        }
    }

    // ================================================================================
    // Dependency Management
    // ================================================================================

    /**
     * Adds a dependency (this issue depends on another).
     *
     * @param command Command containing issue and dependency IDs
     * @return IssueDto representing the updated issue
     * @throws io.spiralhouse.cycletime.application.exceptions.CircularDependencyException if circular dependency detected
     * @throws io.spiralhouse.cycletime.application.exceptions.IssueNotFoundException if either issue doesn't exist
     */
    suspend fun addDependency(command: AddDependencyCommand): IssueDto {
        return unitOfWork.execute {
            val issue = findIssueOrThrow(command.issueId)

            // Validate dependency exists
            val dependency = findIssueOrThrow(command.dependencyId)

            // Check for circular dependency BEFORE adding
            // Note: Domain entity will handle self-dependency validation
            if (command.issueId != command.dependencyId && hasCircularDependency(command.issueId, command.dependencyId)) {
                throw CircularDependencyException(command.issueId, command.dependencyId)
            }

            // Add dependency - domain entity validates (including self-dependency)
            issue.addDependency(command.dependencyId)

            issueRepository.save(issue)
            IssueDto.fromIssue(issue)
        }
    }

    /**
     * Removes a dependency.
     *
     * @param command Command containing issue and dependency IDs
     * @return IssueDto representing the updated issue
     * @throws io.spiralhouse.cycletime.application.exceptions.IssueNotFoundException if the issue doesn't exist
     */
    suspend fun removeDependency(command: RemoveDependencyCommand): IssueDto {
        return unitOfWork.execute {
            val issue = findIssueOrThrow(command.issueId)

            issue.removeDependency(command.dependencyId)
            issueRepository.save(issue)
            IssueDto.fromIssue(issue)
        }
    }

    /**
     * Adds a blocker (another issue blocks this one).
     *
     * @param command Command containing issue and blocker IDs
     * @return IssueDto representing the updated issue
     * @throws io.spiralhouse.cycletime.application.exceptions.IssueNotFoundException if either issue doesn't exist
     */
    suspend fun addBlocker(command: AddBlockerCommand): IssueDto {
        return unitOfWork.execute {
            val issue = findIssueOrThrow(command.issueId)

            // Validate blocker exists
            val blocker = findIssueOrThrow(command.blockerId)

            issue.addBlockedBy(command.blockerId)
            issueRepository.save(issue)
            IssueDto.fromIssue(issue)
        }
    }

    /**
     * Removes a blocker.
     *
     * @param command Command containing issue and blocker IDs
     * @return IssueDto representing the updated issue
     * @throws io.spiralhouse.cycletime.application.exceptions.IssueNotFoundException if the issue doesn't exist
     */
    suspend fun removeBlocker(command: RemoveBlockerCommand): IssueDto {
        return unitOfWork.execute {
            val issue = findIssueOrThrow(command.issueId)

            issue.removeBlockedBy(command.blockerId)
            issueRepository.save(issue)
            IssueDto.fromIssue(issue)
        }
    }

    // ================================================================================
    // Status Operations
    // ================================================================================

    /**
     * Updates issue status with validation.
     *
     * @param command Command containing issue ID and new status
     * @return IssueDto representing the updated issue
     * @throws io.spiralhouse.cycletime.application.exceptions.InvalidStatusTransitionException if transition is invalid
     * @throws io.spiralhouse.cycletime.application.exceptions.IssueNotFoundException if the issue doesn't exist
     */
    suspend fun updateStatus(command: UpdateIssueStatusCommand): IssueDto {
        return unitOfWork.execute {
            val issue = findIssueOrThrow(command.issueId)

            try {
                issue.updateStatus(command.newStatus)
            } catch (e: DomainException) {
                throw InvalidStatusTransitionException(issue.status, command.newStatus, e)
            }

            issueRepository.save(issue)
            IssueDto.fromIssue(issue)
        }
    }

    /**
     * Gets all issues with a specific status.
     *
     * @param status The status to filter by
     * @return List of IssueDto matching the status
     */
    suspend fun getIssuesByStatus(status: IssueStatus): List<IssueDto> {
        return unitOfWork.execute {
            val issues = issueRepository.findByStatus(status)
            issues.map { issue -> IssueDto.fromIssue(issue) }
        }
    }

    // ================================================================================
    // Query Operations
    // ================================================================================

    /**
     * Retrieves all issues in the system.
     *
     * ## Implementation Note:
     * Uses separate queries by type rather than a single findAll() query to maintain
     * consistency with repository interface design. Each type query is optimized
     * separately and the results are combined in memory.
     *
     * @return IssueListDto containing all issues and total count
     */
    suspend fun listIssues(): IssueListDto {
        return unitOfWork.execute {
            // Query each issue type separately and combine results
            val allIssues = listOf(
                issueRepository.findByType(IssueType.EPIC),
                issueRepository.findByType(IssueType.STORY),
                issueRepository.findByType(IssueType.SUBTASK)
            ).flatten()
            
            IssueListDto.fromIssues(allIssues)
        }
    }

    /**
     * Gets all issues associated with a specific project.
     *
     * @param projectId The project ID to filter by
     * @return List of IssueDto associated with the project
     */
    suspend fun getIssuesByProject(projectId: ProjectId): List<IssueDto> {
        return unitOfWork.execute {
            val issues = issueRepository.findByProject(projectId)
            issues.map { issue -> IssueDto.fromIssue(issue) }
        }
    }

    /**
     * Gets all issues assigned to a specific user.
     *
     * @param assigneeId The assignee ID to filter by
     * @return List of IssueDto assigned to the user
     */
    suspend fun getIssuesByAssignee(assigneeId: String): List<IssueDto> {
        return unitOfWork.execute {
            val issues = issueRepository.findByAssignee(assigneeId)
            issues.map { issue -> IssueDto.fromIssue(issue) }
        }
    }

    /**
     * Gets all issues of a specific type.
     *
     * @param type The issue type to filter by
     * @return List of IssueDto of the specified type
     */
    suspend fun getIssuesByType(type: IssueType): List<IssueDto> {
        return unitOfWork.execute {
            val issues = issueRepository.findByType(type)
            issues.map { issue -> IssueDto.fromIssue(issue) }
        }
    }

    // ================================================================================
    // Private Helper Methods
    // ================================================================================

    /**
     * Finds an issue by ID or throws IssueNotFoundException.
     * Reduces duplication of the common find-or-throw pattern.
     *
     * @param issueId The ID of the issue to find
     * @return The found Issue
     * @throws IssueNotFoundException if the issue doesn't exist
     */
    private suspend fun findIssueOrThrow(issueId: IssueId): Issue {
        return issueRepository.findById(issueId)
            ?: throw IssueNotFoundException(issueId)
    }

    /**
     * Validates hierarchy rules between child and parent issue types.
     *
     * ## Hierarchy Rules:
     * - **Epic**: Cannot have any parent (top-level only)
     * - **Story**: Can have Epic parent or no parent (orphaned stories allowed)
     * - **Subtask**: MUST have Story parent (orphaned subtasks NOT allowed)
     *
     * Note: The application layer enforces strict hierarchy for Subtasks to maintain
     * data integrity and prevent orphaned work items. Stories can be orphaned to allow
     * flexible workflow organization. Invalid parent types are always rejected
     * (e.g., Epic with Epic parent, Story with Subtask parent).
     *
     * @param childType The type of the child issue
     * @param parentType The type of the parent issue (null if no parent)
     * @throws HierarchyViolationException if the hierarchy rules are violated
     */
    private fun validateHierarchy(childType: IssueType, parentType: IssueType?) {
        val isValid = when (childType) {
            IssueType.EPIC -> {
                // Epics cannot have any parent
                parentType == null
            }
            IssueType.STORY -> {
                // Stories can have no parent or an Epic parent only
                parentType == null || parentType == IssueType.EPIC
            }
            IssueType.SUBTASK -> {
                // Subtasks MUST have a Story parent (not null, not Epic)
                parentType == IssueType.STORY
            }
        }

        if (!isValid) {
            throw HierarchyViolationException(childType, parentType)
        }
    }

    /**
     * Checks if adding a dependency would create a circular dependency.
     *
     * Uses depth-first search to detect cycles in the dependency graph.
     * Also handles self-dependencies as a special case.
     *
     * @param fromIssueId The issue that would depend on toIssueId
     * @param toIssueId The issue that fromIssueId would depend on
     * @return true if adding this dependency would create a cycle
     */
    private suspend fun hasCircularDependency(fromIssueId: IssueId, toIssueId: IssueId): Boolean {
        // Self-dependency check
        if (fromIssueId == toIssueId) return true

        // DFS to detect circular dependency through the graph
        val visited = mutableSetOf<IssueId>()
        val recursionStack = mutableSetOf<IssueId>()

        suspend fun dfs(currentId: IssueId): Boolean {
            if (currentId == fromIssueId) return true // Found cycle back to start
            if (recursionStack.contains(currentId)) return false // Already in current path
            if (visited.contains(currentId)) return false // Already fully explored

            visited.add(currentId)
            recursionStack.add(currentId)

            val issue = issueRepository.findById(currentId) ?: return false
            for (dependencyId in issue.dependencies) {
                if (dfs(dependencyId)) return true
            }

            recursionStack.remove(currentId)
            return false
        }

        return dfs(toIssueId)
    }
}
